NetworkCaptureLogger: Move SSL logging

sepalani opened the pull request:
This PR moves the SSL logging part to its own logger class. ~~Not sure, if putting it inside SConfig is a good idea so~~ I'm open to suggestions regarding where it should belong.

Ready to be reviewed & merged.

lioncash commented on Source/Core/Core/NetworkCaptureLogger.h:
```diff
@@ -0,0 +1,35 @@
+
+#pragma once
+
+#include "Common/File.h"
```
`<cstddef>` needs to be included for `size_t`.

lioncash commented on Source/Core/Core/NetworkCaptureLogger.h:
```diff
@@ -0,0 +1,35 @@
+
+#include "Common/File.h"
+
+class NetworkCaptureLogger
```
These should likely be in the `Core` namespace.

lioncash commented on Source/Core/Core/NetworkCaptureLogger.cpp:
```diff
@@ -0,0 +1,42 @@
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "NetworkCaptureLogger.h"
```
`"Core/NetworkCaptureLogger.h"`

lioncash commented on Source/Core/Core/NetworkCaptureLogger.cpp:
```diff
@@ -0,0 +1,45 @@
+{
+  m_file_write.WriteBytes(data, length);
+}
+}
```
`}  // namespace Core`

lioncash commented on Source/Core/Core/NetworkCaptureLogger.h:
```diff
@@ -0,0 +1,40 @@
+  File::IOFile m_file_read;
+  File::IOFile m_file_write;
+};
+}
```
`}  // namespace Core`

delroth commented:
@dolphin-emu-bot rebuild

sepalani commented:
Rebased on latest master.

Still ready to be reviewed & merged.

sepalani commented:
Rebased on latest master.

sepalani commented:
Rebased on #8812.

I slightly changed the logger interface and moved the logger to `PowerPC::debug_interface` rather than `SConfig`.

sepalani commented:
Rebased on latest master, I deleted the copy/move constructors and assignment operators as suggested https://github.com/dolphin-emu/dolphin/pull/6075#discussion_r436287711. I also addressed the following comments: https://github.com/dolphin-emu/dolphin/pull/6075#discussion_r436287772 & https://github.com/dolphin-emu/dolphin/pull/6075#discussion_r436287056.

Still ready to be reviewed & merged.

BhaaLseN commented on Source/Core/Core/Debugger/PPCDebugInterface.cpp:
```diff
@@ -382,10 +383,35 @@ void PPCDebugInterface::RunToBreakpoint()
+      m_network_logger = std::make_shared<Core::BinarySSLCaptureLogger>();
+      break;
+    case Core::NetworkCaptureType::None:
+      m_network_logger = std::make_shared<Core::DummyNetworkCaptureLogger>();
```
Did you intentionally leave off the final `break` here?

sepalani replied:
Yes, since it's the last case statement. I can add it back if it's preferred.

BhaaLseN replied:
I don't quite see why that is a reason to omit it; I'd prefer (and I think even _we_ prefer, according to the coding guidelines) that every case stament is terminated with either a `break` or an intentional `[[fallthrough]]`. Makes the code look the same everywhere, and isn't anything to worry about if you add more cases or shuffle them around.

sepalani replied:
Done.

sepalani commented:
Rebased on latest master.

Still ready to be reviewed & merged.

iwubcode commented on Source/Core/Core/NetworkCaptureLogger.h:
```diff
@@ -0,0 +1,48 @@
+  virtual NetworkCaptureType GetCaptureType() const = 0;
+};
+
+class DummyNetworkCaptureLogger : public NetworkCaptureLogger
```
Did you intentionally leave `final` off of this one?

sepalani replied:
Yes, this one is intentional as it will be used to stub unimplemented functions for the `BinarySSLCaptureLogger` class in the follow up PR https://github.com/dolphin-emu/dolphin/pull/6075.

iwubcode commented on Source/Core/Core/Debugger/PPCDebugInterface.cpp:
```diff
@@ -383,10 +384,36 @@ void PPCDebugInterface::RunToBreakpoint()
+{
+  const bool has_ssl = Config::Get(Config::MAIN_NETWORK_SSL_DUMP_READ) ||
+                       Config::Get(Config::MAIN_NETWORK_SSL_DUMP_WRITE);
+  const auto current_capture_type =
```
It's possible I'm missing something but I feel like this might be a little bit simpler:

```
if (has_ssl)
{
  if (!m_network_logger || m_network_logger->GetCaptureType() != Core::NetworkCaptureType::Raw)
    m_network_logger = std::make_shared<Core::BinarySSLCaptureLogger>();
 }
 else
{
  if (!m_network_logger || m_network_logger->GetCaptureType() != Core::NetworkCaptureType::None)
    m_network_logger = std::make_shared<Core::DummyNetworkCaptureLogger>();
}

return m_network_logger;
```

sepalani replied:
This function acts a bit like a singleton pattern, if the logger isn't initialized or has been changed it creates a new one.

The alternative you provided doesn't sound very scalable to me if you have more logging types (ex: the PCAP one will be in a follow up PR). Your alternative also call `GetCaptureType()` multiple times.

iwubcode replied:
Ah ok, sorry, I didn't realize there were more types planned.  I'm curious though, how will that work with the current config.  There's only a single boolean for ssl at the moment.

I see what you're trying to go for but I personally feel like the `reset()` then call `make_shared` in the same pass, along with multiple checks to do so is quite odd.  That's why I was trying to suggest an alternative.

Maybe a simple change would suffice:

```
if (m_network_logger && m_network_logger->GetCaptureType() == current_capture_type)
  return m_network_logger;

switch (current_capture_type)
{
  case Core::NetworkCaptureType::Raw:
    m_network_logger = std::make_shared<Core::BinarySSLCaptureLogger>();
    break;
  case Core::NetworkCaptureType::None:
    m_network_logger = std::make_shared<Core::DummyNetworkCaptureLogger>();
    break;
};

return m_network_logger;
```

Thoughts?

sepalani replied:
Yes, you have a point. Inverting the condition is way easier to read. I'm going to modify that, ASAP.

Dentomologist commented on Source/Core/Core/Debugger/PPCDebugInterface.cpp:
```diff
@@ -383,10 +384,36 @@ void PPCDebugInterface::RunToBreakpoint()
 {
 }
 
+std::shared_ptr<Core::NetworkCaptureLogger> PPCDebugInterface::NetworkLogger()
+{
+  const bool has_ssl = Config::Get(Config::MAIN_NETWORK_SSL_DUMP_READ) ||
+                       Config::Get(Config::MAIN_NETWORK_SSL_DUMP_WRITE);
+  const auto current_capture_type =
+      has_ssl ? Core::NetworkCaptureType::Raw : Core::NetworkCaptureType::None;
+
+  if (m_network_logger && m_network_logger->GetCaptureType() != current_capture_type)
+    m_network_logger.reset();
+
+  if (!m_network_logger)
+  {
+    switch (current_capture_type)
+    {
+    case Core::NetworkCaptureType::Raw:
+      m_network_logger = std::make_shared<Core::BinarySSLCaptureLogger>();
+      break;
+    case Core::NetworkCaptureType::None:
+      m_network_logger = std::make_shared<Core::DummyNetworkCaptureLogger>();
+      break;
+    }
+  }
+  return m_network_logger;
+}
```
How do you feel about something along these lines?

```
namespace {
std::shared_ptr<Core::NetworkCaptureLogger> CreateNetworkCaptureLogger(const Core::NetworkCaptureType capture_type) {
	switch(capture_type) {
	case Core::NetworkCaptureType::Raw:
		return std::make_shared<Core::BinarySSLCaptureLogger>();
	case Core::NetworkCaptureType::None:
		return std::make_shared<Core::DummyNetworkCaptureLogger>();	
	}
}
}  // namespace

bool PPCDebugInterface::LoggerHasCaptureType(const Core::NetworkCaptureType capture_type) const {
	return m_network_logger && m_network_logger->GetCaptureType() == capture_type;
}

std::shared_ptr<Core::NetworkCaptureLogger> PPCDebugInterface::NetworkLogger()
{
	const bool has_ssl = Config::Get(Config::MAIN_NETWORK_SSL_DUMP_READ) ||
		Config::Get(Config::MAIN_NETWORK_SSL_DUMP_WRITE);
	const auto current_capture_type = has_ssl ? Core::NetworkCaptureType::Raw : Core::NetworkCaptureType::None;

	if (!LoggerHasCaptureType(current_capture_type))
	{
		m_network_logger = CreateNetworkCaptureLogger(current_capture_type);
	}
	return m_network_logger;
}
```

(Alternately you could assign m_network_logger directly in CreateNetworkCaptureLogger).

Unless I've misunderstood how shared_ptr works you shouldn't need to call reset() before assigning a new value to it, so we don't need to special case that.

iwubcode replied:
@Dentomologist - that's basically what I came up with in my comment.  While I thought about it, I didn't bother with the helper functions.  If the logic will be used in multiple places, it might make sense to use helpers but the `NetworkLogger` function is small enough I don't think it's necessary and actually means you have to look somewhere else to know what is happening.  Still, I'd be fine whichever @sepalani chooses!

iwubcode reviewed (approved):
LGTM (untested)

iwubcode commented on Source/Core/Core/Core.cpp:
```diff
@@ -534,6 +534,7 @@ static void EmuThread(std::unique_ptr<BootParameters> boot, WindowSystemInfo wsi
 
     PatchEngine::Shutdown();
     HLE::Clear();
+    PowerPC::debug_interface.Clear();
```
Right after I hit the approve button, I realized I glossed over this.  What's this for?

sepalani replied:
It clears the debug_interface after a full shutdown (i.e. breakpoints, patches, watches, memcheck) and it also releases the network logger and its file.

I have the feeling it's an artifact that wasn't migrated when switching to Qt, I might be wrong but I don't recall adding such line when I originally created the PR.

iwubcode replied:
Ah ok, looks like this might also resolve #9325 and the issue that was mentioned there?

sepalani replied:
Sorry, my wording was poorly chosen. By "full shutdown" I meant the game/Wii shutdown not the actual game so it won't. 

After a closer look at the current code, I'll probably add a 2nd commit that removes the GUI code that takes care of the breakpoints/memchecks clearing as it's using Qt signals and Dolphin state rather than removing them on shutdown in core.

sepalani replied:
Unnecessary GUI code removed in the 2nd commit.

leoetlino commented on Source/Core/Core/Debugger/PPCDebugInterface.h:
```diff
@@ -82,9 +83,12 @@ class PPCDebugInterface final : public Common::DebugInterface
   u32 GetColor(u32 address) const override;
   std::string GetDescription(u32 address) const override;
 
+  std::shared_ptr<Core::NetworkCaptureLogger> NetworkLogger();
```
`#include <memory>`

sepalani replied:
Done.

leoetlino commented on Source/Core/Core/NetworkCaptureLogger.cpp:
```diff
@@ -0,0 +1,52 @@
+
+void BinarySSLCaptureLogger::LogRead(const void* data, std::size_t length)
+{
+  if (!Config::Get(Config::MAIN_NETWORK_SSL_DUMP_READ))
```
Is this check actually useful, considering you're already checking those config settings in the logger instance getter?

leoetlino commented on Source/Core/Core/NetworkCaptureLogger.cpp:
```diff
@@ -0,0 +1,52 @@
+
+void BinarySSLCaptureLogger::LogWrite(const void* data, std::size_t length)
+{
+  if (!Config::Get(Config::MAIN_NETWORK_SSL_DUMP_WRITE))
```
Same question here.

sepalani replied:
Ditto but for SSL write.
